File: packages/ios/Sources/Components/FloorSelector/FloorSelector.swift
import SwiftUI

/// Layout of the floor selector, mirroring the React `FloorSelector.variant`.
public enum KozmosFloorSelectorVariant: String, CaseIterable, Sendable {
    case verticalList = "vertical-list"
    case horizontalList = "horizontal-list"
    case compactStepper = "compact-stepper"
    /// Shows the current level only, and opens the full list when touched.
    /// For a control parked in a corner of a map, where a permanent column of
    /// every level costs more of the map than it is worth.
    case collapsible = "collapsible"
}

/// Switches the active level of a venue.
///
/// Mirrors the React `FloorSelector` API. `selectedFloor` is the canonical
/// floor ID, never a display label, and the component reports selection rather
/// than deriving it.
///
/// Pass `[KozmosFloorPresentation]` so the button shows the level's
/// `shortLabel` while selection and analytics stay keyed on its `id`, and so
/// assistive technology hears the full `label`. The `[String]` initializer is
/// for venues whose IDs are already the labels — it shows each ID as-is.
public struct KozmosFloorSelector: View {
    let floors: [KozmosFloorPresentation]
    @Binding var selectedFloor: String
    let variant: KozmosFloorSelectorVariant
    let label: String
    @Environment(\.kozmosAnalytics) private var trackEvent

    /// A fixed 40pt button truncates every level to an ellipsis once Dynamic
    /// Type is turned up, which leaves the control unreadable — and a floor
    /// selector whose floors cannot be told apart is not a control at all. The
    /// target scales with the type it has to hold.
    @ScaledMetric(relativeTo: .subheadline)
    private var controlSize: CGFloat = KozmosDimensions.primitivesLayoutSizing500

    @State private var isExpanded = false

    public init(
        floors: [KozmosFloorPresentation],
        selectedFloor: Binding<String>,
        variant: KozmosFloorSelectorVariant = .verticalList,
        label: String = "Floor selector"
    ) {
        self.floors = floors
        self._selectedFloor = selectedFloor
        self.variant = variant
        self.label = label
    }

    /// For tests and previews: the collapsible list already open.
    init(
        floors: [KozmosFloorPresentation],
        selectedFloor: Binding<String>,
        variant: KozmosFloorSelectorVariant,
        label: String = "Floor selector",
        expanded: Bool
    ) {
        self.init(floors: floors, selectedFloor: selectedFloor, variant: variant, label: label)
        self._isExpanded = State(initialValue: expanded)
    }

    /// For venues with no separate display label: the ID is rendered verbatim.
    public init(
        floors: [String],
        selectedFloor: Binding<String>,
        variant: KozmosFloorSelectorVariant = .verticalList,
        label: String = "Floor selector"
    ) {
        self.init(
            floors: floors.map {
                KozmosFloorPresentation(id: $0, label: $0, shortLabel: $0)
            },
            selectedFloor: selectedFloor,
            variant: variant,
            label: label
        )
    }

    private func select(_ floor: KozmosFloorPresentation) {
        guard !floor.disabled else { return }
        if variant == .collapsible {
            withAnimation(.spring(response: 0.3, dampingFraction: 0.85)) { isExpanded = false }
        }
        trackEvent(
            KozmosAnalyticsEvent(
                eventName: "floor_selected",
                component: "FloorSelector",
                properties: ["floor": floor.id]
            )
        )
        withAnimation { selectedFloor = floor.id }
    }

    var selectedIndex: Int {
        floors.firstIndex { $0.id == selectedFloor } ?? 0
    }

    private var selectedPresentation: KozmosFloorPresentation? {
        floors.first { $0.id == selectedFloor } ?? floors.first
    }

    /// While the list is open the closed control would only repeat the level
    /// already highlighted in it, so it is hidden — but it keeps its space, or
    /// the control would change size and move the map after all.
    private var baseIsHidden: Bool {
        variant == .collapsible && isExpanded
    }

    public var body: some View {
        container
            .opacity(baseIsHidden ? 0 : 1)
            .padding(KozmosDimensions.primitivesLayoutSpacing75)
            .background(baseIsHidden ? Color.clear : KozmosColors.primitivesColorsBackground0.opacity(0.9))
            .cornerRadius(KozmosDimensions.semanticsRadiusPanel)
            .kozmosElevation(baseIsHidden ? KozmosShadows.none : KozmosShadows.semanticsElevationFloating)
            // Overlaid rather than stacked, so opening the list does not change
            // what this control measures. A map shell reports the space its
            // chrome covers, and a camera that re-frames every time a picker
            // opens is worse than one that ignores it. Trailing-aligned: the
            // list is wider than the control it opens from, and grows away
            // from the map's edge the control sits at.
            .overlay(alignment: .bottomTrailing) { expandedList }
            .accessibilityElement(children: .contain)
            .accessibilityLabel(label)
    }

    /// The full list, floating above the closed control: every level's short
    /// label in its square with the level's name beside it, the current one
    /// filled. A column of "L1, L2" alone told a visitor nothing they could
    /// not read off the closed pill.
    @ViewBuilder
    private var expandedList: some View {
        if variant == .collapsible, isExpanded {
            VStack(alignment: .leading, spacing: KozmosDimensions.primitivesLayoutSpacing100) {
                ForEach(floors) { floor in
                    namedFloorButton(floor)
                }
            }
            .padding(KozmosDimensions.primitivesLayoutSpacing75)
            .background(KozmosColors.primitivesColorsBackground0.opacity(0.9))
            .cornerRadius(KozmosDimensions.semanticsRadiusPanel)
            .kozmosElevation(KozmosShadows.semanticsElevationFloating)
            .fixedSize()
            .offset(
                y: -(controlSize
                     + KozmosDimensions.primitivesLayoutSpacing75 * 2
                     + KozmosDimensions.primitivesLayoutSpacing100)
            )
            .transition(.opacity.combined(with: .scale(scale: 0.92, anchor: .bottom)))
        }
    }

    @ViewBuilder
    private var container: some View {
        switch variant {
        case .verticalList:
            VStack(spacing: KozmosDimensions.primitivesLayoutSpacing100) {
                ForEach(floors) { floor in
                    floorButton(floor)
                }
            }
        case .horizontalList:
            HStack(spacing: KozmosDimensions.primitivesLayoutSpacing100) {
                ForEach(floors) { floor in
                    floorButton(floor)
                }
            }
        case .compactStepper:
            VStack(spacing: KozmosDimensions.primitivesLayoutSpacing100) {
                stepperButton(systemImage: "chevron.up", step: -1, label: "Floor up")
                if let selectedPresentation {
                    floorButton(selectedPresentation)
                }
                stepperButton(systemImage: "chevron.down", step: 1, label: "Floor down")
            }
        case .collapsible:
            // Only ever the current level. The list that opens is an overlay,
            // not part of this footprint — see `expandedList`.
            if let selectedPresentation {
                collapsedButton(selectedPresentation)
            }
        }
    }

    private func floorButton(_ floor: KozmosFloorPresentation) -> some View {
        let isSelected = floor.id == selectedFloor
        return Button {
            select(floor)
        } label: {
            Text(floor.shortLabel)
                .font(KozmosTypography.subheadline)
                .bold()
                .lineLimit(1)
                .minimumScaleFactor(0.7)
                .frame(width: controlSize, height: controlSize)
                .background(isSelected ? KozmosColors.primitivesColorsTheme500 : Color.clear)
                .foregroundColor(
                    isSelected
                        ? KozmosColors.primitivesColorsBackground0
                        : KozmosColors.primitivesColorsForeground100
                )
                .cornerRadius(KozmosDimensions.semanticsRadiusPanel)
                // An unselected button is transparent, so without an explicit
                // hit shape only the glyph itself would accept a tap.
                .contentShape(
                    RoundedRectangle(
                        cornerRadius: KozmosDimensions.semanticsRadiusPanel,
                        style: .continuous
                    )
                )
        }
        .buttonStyle(.plain)
        .disabled(floor.disabled)
        .opacity(floor.disabled ? 0.4 : 1)
        // The button shows the short label; assistive technology gets the full
        // one, which is the only place the level is spelled out.
        .accessibilityLabel(floor.label)
        .accessibilityAddTraits(isSelected ? [.isButton, .isSelected] : .isButton)
    }

    /// A row of the open list: the short label in its square, the name beside
    /// it when the venue gives one. One button, so a tap anywhere on the row
    /// selects, and assistive technology hears the name once.
    private func namedFloorButton(_ floor: KozmosFloorPresentation) -> some View {
        let isSelected = floor.id == selectedFloor
        return Button {
            select(floor)
        } label: {
            HStack(spacing: KozmosDimensions.primitivesLayoutSpacing100) {
                Text(floor.shortLabel)
                    .font(KozmosTypography.subheadline)
                    .bold()
                    .lineLimit(1)
                    .minimumScaleFactor(0.7)
                    .frame(width: controlSize, height: controlSize)
                    .background(isSelected ? KozmosColors.primitivesColorsTheme500 : Color.clear)
                    .foregroundColor(
                        isSelected
                            ? KozmosColors.primitivesColorsBackground0
                            : KozmosColors.primitivesColorsForeground100
                    )
                    .cornerRadius(KozmosDimensions.semanticsRadiusPanel)
                if floor.label != floor.shortLabel {
                    Text(floor.label)
                        .font(KozmosTypography.subheadline)
                        .fontWeight(isSelected ? .semibold : .regular)
                        .foregroundColor(KozmosColors.primitivesColorsForeground100)
                        .lineLimit(1)
                        .padding(.trailing, KozmosDimensions.primitivesLayoutSpacing100)
                }
            }
            .contentShape(
                RoundedRectangle(
                    cornerRadius: KozmosDimensions.semanticsRadiusPanel,
                    style: .continuous
                )
            )
        }
        .buttonStyle(.plain)
        .disabled(floor.disabled)
        .opacity(floor.disabled ? 0.4 : 1)
        .accessibilityLabel(floor.label)
        .accessibilityAddTraits(isSelected ? [.isButton, .isSelected] : .isButton)
    }

    /// The closed state: the level you are on, and a way in to the rest.
    private func collapsedButton(_ floor: KozmosFloorPresentation) -> some View {
        Button {
            trackEvent(
                KozmosAnalyticsEvent(
                    eventName: "floor_selector_expanded",
                    component: "FloorSelector",
                    properties: ["floor": floor.id]
                )
            )
            withAnimation(.spring(response: 0.3, dampingFraction: 0.85)) { isExpanded = true }
        } label: {
            Text(floor.shortLabel)
                .font(KozmosTypography.subheadline)
                .bold()
                .lineLimit(1)
                .minimumScaleFactor(0.7)
                .frame(width: controlSize, height: controlSize)
                .background(KozmosColors.primitivesColorsTheme500)
                .foregroundColor(KozmosColors.primitivesColorsBackground0)
                .cornerRadius(KozmosDimensions.semanticsRadiusPanel)
                .contentShape(
                    RoundedRectangle(
                        cornerRadius: KozmosDimensions.semanticsRadiusPanel,
                        style: .continuous
                    )
                )
        }
        .buttonStyle(.plain)
        .accessibilityLabel(floor.label)
        .accessibilityHint("Shows every level")
        .accessibilityAddTraits(.isButton)
    }

    /// The next selectable floor in list order, skipping any that are closed.
    func reachableIndex(step: Int) -> Int? {
        guard !floors.isEmpty else { return nil }
        var candidate = selectedIndex + step
        while floors.indices.contains(candidate) {
            if !floors[candidate].disabled { return candidate }
            candidate += step
        }
        return nil
    }

    /// Steps through the floor list. `step` is in list order, so -1 is the
    /// entry above the current one.
    private func stepperButton(systemImage: String, step: Int, label: String) -> some View {
        let target = reachableIndex(step: step)

        return Button {
            if let target { select(floors[target]) }
        } label: {
            Image(systemName: systemImage)
                .font(.system(size: 14, weight: .bold))
                .frame(width: controlSize, height: controlSize)
                .foregroundColor(KozmosColors.primitivesColorsForeground500)
                .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        .disabled(target == nil)
        .opacity(target == nil ? 0.4 : 1)
        .accessibilityLabel(label)
    }
}

File: packages/android/src/main/java/com/kozmos/components/FloorSelector/FloorSelector.kt
package com.kozmos.components.floorselector

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.KeyboardArrowDown
import androidx.compose.material.icons.filled.KeyboardArrowUp
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.selected
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.kozmos.providers.KozmosAnalyticsEvent
import com.kozmos.providers.LocalKozmosAnalytics
import com.kozmos.tokens.KozmosShadows
import com.kozmos.tokens.KozmosColors
import com.kozmos.tokens.KozmosDimensions

/** Layout of the floor selector, mirroring the React `FloorSelector.variant`. */
enum class KozmosFloorSelectorVariant(val value: String) {
    VerticalList("vertical-list"),
    HorizontalList("horizontal-list"),
    CompactStepper("compact-stepper")
}

/**
 * Switches the active level of a venue.
 *
 * Mirrors the React `FloorSelector` API. [selectedFloor] is the canonical floor
 * ID, never a display label, and the component reports selection rather than
 * deriving it.
 */
@Composable
fun KozmosFloorSelector(
    floors: List<String>,
    selectedFloor: String,
    onFloorSelect: (String) -> Unit,
    modifier: Modifier = Modifier,
    variant: KozmosFloorSelectorVariant = KozmosFloorSelectorVariant.VerticalList,
    label: String = "Floor selector"
) {
    val trackEvent = LocalKozmosAnalytics.current
    val selectedIndex = floors.indexOf(selectedFloor).takeIf { it >= 0 } ?: 0

    fun select(floor: String) {
        trackEvent(
            KozmosAnalyticsEvent(
                component = "FloorSelector",
                eventName = "floor_selected",
                properties = mapOf("floor" to floor)
            )
        )
        onFloorSelect(floor)
    }

    val floorButton: @Composable (String) -> Unit = { floor ->
        val isSelected = floor == selectedFloor
        Box(
            modifier = Modifier
                .size(KozmosDimensions.primitivesLayoutSizing500)
                .clip(RoundedCornerShape(KozmosDimensions.semanticsRadiusPanel))
                .background(
                    if (isSelected) KozmosColors.primitivesColorsTheme500 else Color.Transparent
                )
                .clickable { select(floor) }
                .semantics {
                    contentDescription = floor
                    selected = isSelected
                },
            contentAlignment = Alignment.Center
        ) {
            Text(
                text = floor,
                color = if (isSelected) {
                    KozmosColors.primitivesColorsBackground0
                } else {
                    KozmosColors.primitivesColorsForeground100
                },
                style = MaterialTheme.typography.labelLarge.copy(fontWeight = FontWeight.Bold)
            )
        }
    }

    // Steps through the floor list. `offset` is in list order, so -1 is the
    // entry above the current one.
    val stepperButton: @Composable (Int, String) -> Unit = { offset, description ->
        val target = selectedIndex + offset
        val isEnabled = target in floors.indices
        Box(
            modifier = Modifier
                .size(KozmosDimensions.primitivesLayoutSizing500)
                .alpha(if (isEnabled) 1f else 0.4f)
                .clickable(enabled = isEnabled) { select(floors[target]) }
                .semantics { contentDescription = description },
            contentAlignment = Alignment.Center
        ) {
            Icon(
                imageVector = if (offset < 0) {
                    Icons.Default.KeyboardArrowUp
                } else {
                    Icons.Default.KeyboardArrowDown
                },
                contentDescription = null,
                tint = KozmosColors.primitivesColorsForeground500
            )
        }
    }

    val rootModifier = modifier
        .shadow(KozmosShadows.semanticsElevationFloating, RoundedCornerShape(KozmosDimensions.semanticsRadiusPanel))
        .background(
            KozmosColors.primitivesColorsBackground0,
            RoundedCornerShape(KozmosDimensions.semanticsRadiusPanel)
        )
        .padding(KozmosDimensions.primitivesLayoutSpacing50)
        .semantics { contentDescription = label }

    val spacing = Arrangement.spacedBy(KozmosDimensions.primitivesLayoutSpacing100)

    when (variant) {
        KozmosFloorSelectorVariant.HorizontalList -> Row(
            modifier = rootModifier,
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = spacing
        ) {
            floors.forEach { floorButton(it) }
        }

        KozmosFloorSelectorVariant.VerticalList -> Column(
            modifier = rootModifier,
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = spacing
        ) {
            floors.forEach { floorButton(it) }
        }

        KozmosFloorSelectorVariant.CompactStepper -> Column(
            modifier = rootModifier,
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = spacing
        ) {
            stepperButton(-1, "Floor up")
            floorButton(selectedFloor)
            stepperButton(1, "Floor down")
        }
    }
}
